include/nbl/video/IGPUCommandPool.h

```diff
@@ -4,38 +4,803 @@
+                // get reclaimed in ~LinearAddressAllocator
+
+                // We potentially continue to the next command segment under any one of the two conditions:
+                const bool potentiallyContinueToNextSegment =
+                    // 1. If the we run past the storage of the current segment.
+                    ((reinterpret_cast<uint8_t*>(itr.m_cmd) - itr.m_segment->getData()) >= IGPUCommandPool::CCommandSegment::STORAGE_SIZE)
+                    ||
+                    // 2. If we encounter a 0-sized command (terminating command) before running out of the current segment. This case will arise when the current
+                    // segment doesn't have enough storage to hold the next command.
+                    (itr.m_cmd->getSize() == 0);
+
+                if (potentiallyContinueToNextSegment)
+                {
+                    CCommandSegment* nextSegment = currSegment->getNext();
+
+                    currSegment->~CCommandSegment();
+                    m_commandSegmentPool.deallocate(currSegment, COMMAND_SEGMENT_SIZE);
+
+                    if (!nextSegment)
+                        break;
+
+                    itr.m_segment = nextSegment;
+                    itr.m_cmd = itr.m_segment->getFirstCommand();
+                }
+
+                lastCmd = itr.m_cmd->getSize() == 0u;
+                if (lastCmd)
+                {
+                    currSegment->~CCommandSegment();
+                    m_commandSegmentPool.deallocate(currSegment, COMMAND_SEGMENT_SIZE);
+                }
+            }
+
+            segmentListHeadItr.m_cmd = nullptr;
+            segmentListHeadItr.m_segment = nullptr;
+            segmentListTail = nullptr;
+        }
+    }
+
+    bool reset()
+    {
+        m_resetCount.fetch_add(1);
+        // TODO(achal): Reset the entire m_commandSegmentPool without keeping track of the command buffer's heads.
+        return reset_impl();
+    }
+
+    inline uint32_t getResetCounter() { return m_resetCount.load(); }
+
+protected:
+    virtual ~IGPUCommandPool() = default;
+
+    virtual bool reset_impl() { return true; };
+
+    core::bitflag<E_CREATE_FLAGS> m_flags;
+    uint32_t m_familyIx;
+
+private:
+    std::atomic_uint32_t m_resetCount = 0;
+    core::CMemoryPool<core::PoolAddressAllocator<uint32_t>, core::default_aligned_allocator, false, uint32_t> m_commandSegmentPool;
+};
+
+class IGPUCommandPool::CBeginCmd : public IGPUCommandPool::IFixedSizeCommand<CBeginCmd>
+{
+public:
+    CBeginCmd(core::smart_refctd_ptr<const IGPURenderpass>&& renderpass, core::smart_refctd_ptr<const IGPUFramebuffer>&& framebuffer) : m_renderpass(std::move(renderpass)), m_framebuffer(std::move(framebuffer)) {}
+
+private:
+    core::smart_refctd_ptr<const IGPURenderpass> m_renderpass;
+    core::smart_refctd_ptr<const IGPUFramebuffer> m_framebuffer;
+};
```

Comment:
what is this command?

Reply:
This gets emplaced in `IGPUCommandBuffer::begin` when the command buffer has `SInheritanceInfo` because I needed a way to keep the renderpass and framebuffer resources alive.